libraries/SPI/SPI.cpp

```diff
@@ -531,48 +528,90 @@ void SPIClass::transferBytes(const uint8_t * out, uint8_t * in, uint32_t size) {
 }
 
 /**
- * Note:
- *  in and out need to be aligned to 32Bit
- *  or you get an Fatal exception (9)
  * @param out uint8_t *
  * @param in  uint8_t *
  * @param size uint8_t (max 64)
  */
 void SPIClass::transferBytes_(const uint8_t * out, uint8_t * in, uint8_t size) {
-    while(SPI1CMD & SPIBUSY) {}
-    // Set in/out Bits to transfer
+    while(SPI1CMD & SPIBUSY) {} // Make sure we're IDLE on the SPI
 
-    setDataBits(size * 8);
+    int bytesToTransfer = size; // How much to read/write from the FIFO in bytes
 
-    volatile uint32_t * fifoPtr = &SPI1W0;
-    uint8_t dataSize = ((size + 3) / 4);
+    if (out && !in) {
+        int bytesLeft = size;
 
-    if(out) {
-        uint32_t * dataPtr = (uint32_t*) out;
-        while(dataSize--) {
-            *fifoPtr = *dataPtr;
-            dataPtr++;
-            fifoPtr++;
+        // Only transmitting, get out 32b aligned
+        while (bytesLeft && ((uint32_t)out & 3)) {
+            SPI.transfer(*(out++));
+            bytesLeft--;
         }
-    } else {
-        // no out data only read fill with dummy data!
-        while(dataSize--) {
-            *fifoPtr = 0xFFFFFFFF;
-            fifoPtr++;
+
+        if (!bytesLeft) return;
+        bytesToTransfer = bytesLeft;
+
+        // Now do 32b writes until we have 0 or fewer bytes left
+        // Note we can read and store past the end of string because it will not be sent (setDataBits)
+        uint32_t *dataPtr = (uint32_t*)out;
+        volatile uint32_t *fifoPtr = &SPI1W0;
+        while (bytesLeft > 0) {
+            *(fifoPtr++) = *(dataPtr++);
+            bytesLeft -= 4;
         }
+        // Remainder will be sent out of loaded FIFO below
+    } else if (in && !out) {
+        int bytesLeft = size;
+
+        // Only receiving, get in 32b aligned
+        while (bytesLeft && ((uint32_t)in & 3)) {
+            *(in++) = SPI.transfer(0xff);
```

Comment:
Shouldn't this be just transfer(0xff)?

Reply:
Refactored to remove this bit completely.